psbt/psbt.go

```diff
@@ -0,0 +1,676 @@
+	for _, kvPairList := range pairsList {
+		for _, kvPair := range kvPairList.Items {
+			buf, err := serializeKeyValuePair(kvPair)
+			if err != nil {
+				return err
+			}
+			p.Raw = append(p.Raw, buf.Bytes()...)
+		}
+		p.Raw = append(p.Raw, 0x00)
+	}
+	return nil
+}
+
+// B64Encode returns the base64 encoding of the current
+// serialized Psbt in Psbt.Raw (note that the latter
+// will only be synchronized after a call to .Serialize()),
+// or an error if the encoding fails.
+func (p *Psbt) B64Encode() string {
+	return base64.StdEncoding.EncodeToString(p.Raw)
+}
+
+// IsInputSane returns true only if there are no conflicting
+// values in the Psbt at the input index inIndex. It checks that
+// both witness and non-witness utxo entries do not exist, and that
+// witnessScript entries are only added to witness inputs.
+func (p *Psbt) IsInputSane(inIndex int) bool {
+	var nonWitnessUtxoPresent = false
+	var witnessUtxoPresent = false
+	kvPairList := (*p.InKVPairs)[inIndex]
+	if kvPairList.containsKeyType(PsbtInNonWitnessUtxo) {
+		nonWitnessUtxoPresent = true
+	}
+	if kvPairList.containsKeyType(PsbtInWitnessUtxo) {
+		witnessUtxoPresent = true
+	}
+	if nonWitnessUtxoPresent && witnessUtxoPresent {
+		return false
+	}
+	var witnessScriptPresent = false
+	if kvPairList.containsKeyType(PsbtInWitnessScript) {
+		witnessScriptPresent = true
+	}
+	if witnessScriptPresent && !witnessUtxoPresent {
+		return false
+	}
+	return true
+}
+
+// IsComplete returns true only if all of the inputs are
+// finalized; this is particularly important in that it decides
+// whether the final extraction to a network serialized signed
+// transaction will be possible.
+func (p *Psbt) IsComplete() bool {
+	for i := 0; i < len(p.UnsignedTx.TxIn); i++ {
+		if !isFinalized(p, i) {
+			return false
+		}
+	}
+	return true
+}
+
+// SanityCheck checks conditions on a PSBT to ensure that it obeys the
+// rules of BIP174, and returns true if so, false if not.
+func (p *Psbt) SanityCheck() bool {
+
+	if !validateUnsignedTX(p.UnsignedTx) {
+		return false
+	}
+	err := p.Serialize()
+	if err != nil {
+		return false
+	}
+	for i, _ := range p.UnsignedTx.TxIn {
+		if !p.IsInputSane(i) {
+			return false
+		}
+	}
+	return true
+}
+
+func sectionEmpty(section []PsbtKeyValuePairList) bool {
+	for _, x := range section {
+		if len(x.Items) > 0 {
+			return false
+		}
+	}
+	return true
+}
+func (p *Psbt) InSectionEmpty() bool {
+	return sectionEmpty(*p.InKVPairs)
+}
+
+func (p *Psbt) OutSectionEmpty() bool {
+	return sectionEmpty(*p.OutKVPairs)
+}
+
+// SerializeBIP32Derivation takes a master key fingerprint
+// as defined in BIP32, along with a path specified as a list
+// of uint32 values, and returns a bytestring specifying the derivation
+// in the format required by BIP174:
+// // master key fingerprint (4) || child index (4) || child index (4) || ...
+func SerializeBIP32Derivation(masterKeyFingerprint []byte,
+	bip32Path []uint32) []byte {
+	//TODO fixed length of 4 bytes
```

Comment:
This comment is about the masterKeyFingerprint being variable size? You could [4]byte it, I'd personally use uint32 for this

Reply:
Yes, I went with your suggestion of using uint32, but only because it looks ugly to copy a byte slice to a byte array. Not that it really matters I guess. Done here: b259236475cd8705a63e330a8904070036a349d7